### PR TITLE
support calico

### PR DESCRIPTION
**Pull request by luckymrwang**
Fix the bug which not absolutely compatible with calico [https://github.com/openyurtio/raven/issues/38](https://github.com/openyurtio/raven/issues/38)

### CODE REVIEW

**Comment by njucjc** [comment]
Please deal with `golint` error in CI first.

**Comment by luckymrwang** [comment]
> Please deal with `golint` error in CI first.

The latest code has been commited. `golint` error has been dealed.

**Comment by DrmagicE** [comment]
Thanks for the enhancement. Quite busy these days, I will start to review this Friday.

File: go.mod [hunk 1]
```diff
@@ -3,11 +3,12 @@ module github.com/openyurtio/raven-controller-manager
...
-	github.com/go-logr/logr v0.3.0
+	github.com/go-logr/logr v1.2.0
+	github.com/projectcalico/api v0.0.0-20220505235232-ce7a5122e146
 	github.com/stretchr/testify v1.7.1
-	k8s.io/api v0.19.7
-	k8s.io/apimachinery v0.19.7
-	k8s.io/client-go v0.19.2
-	k8s.io/klog/v2 v2.2.0
-	sigs.k8s.io/controller-runtime v0.7.2
+	k8s.io/api v0.23.2
```
Comment (DrmagicE) [inline go.mod:9]: The latest kubernetes version that OpenYurt currently supports is 1.22 and most of the OpenYurt components use v0.22.3 as kubernetes library version. 

So how about downgrading the version to v0.22.3.

File: pkg/ravencontroller/apis/calico/v3/blockaffinity.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019 Tigera, Inc. All rights reserved.
```
Comment (DrmagicE) [inline pkg/ravencontroller/apis/calico/v3/blockaffinity.go:2]: Please add OpenYurt Copyright to the files in `pkg/ravencontroller/apis/calico/v3/` directory.
For example:
```
// Copyright 2020 The OpenYurt Authors.
// Copyright (c) 2019 Tigera, Inc. All rights reserved.
```

File: pkg/ravencontroller/controllers/gateway_controller.go [hunk 3]
```diff
@@ -238,6 +245,30 @@ func getNodeInternalIP(node corev1.Node) string {
...
+			if err != nil {
+				err = fmt.Errorf("unable to list calico blockaffinity: %s", err)
+				return "", err
+			}
+
+			for _, v := range blockAffinityList.Items {
+				if v.Spec.Node != node.Name || v.Spec.State != "confirmed" {
+					continue
+				}
+				podCIDR = v.Spec.CIDR
```
Comment (DrmagicE) [inline pkg/ravencontroller/controllers/gateway_controller.go:264]: It is possible for nodes in calico to have multiple cidr block, so there can be multiple `blockaffinities` referring the same node.
Currently, the Gateway CR cannot handle multiple podCIDR circumstances well(`NodeInfo.Subnet` is a string).  I think we may add a subnet slice to `Gateway` CR to handle this, and the `raven-agent` also needs some adjustments.

@BSWANG @njucjc any thoughts on this?

Comment (luckymrwang) [reply pkg/ravencontroller/controllers/gateway_controller.go:264]: @DrmagicE You are right. I have realized it too when adding calico cidr block. As you mentioned that `raven-agent` and gateway CR must be adjusted, so I just get one cidr block to adapt current case. When `raven-agent` supported multiple cidr block, I will adjust calico to adapt it again.

File: go.mod (hunk 1 above)
Comment (luckymrwang) [reply go.mod:9]: The latest code has been commited. `k8s components` have been downgraded to v0.22.3.

File: pkg/ravencontroller/apis/calico/v3/blockaffinity.go (hunk 2 above)
Comment (luckymrwang) [reply pkg/ravencontroller/apis/calico/v3/blockaffinity.go:2]: The latest code has been commited. `Copyright 2020 The OpenYurt Authors` have been added.

File: pkg/ravencontroller/controllers/gateway_controller.go (hunk 3 above)
Comment (njucjc) [reply pkg/ravencontroller/controllers/gateway_controller.go:264]: how about changing `NodeInfo.Subnet`(string) to `NodeInfo.Subnets`(slice)，this field is obtained by the controller and will not affect the way users use it.

File: go.mod (hunk 1 above)
Comment (DrmagicE) [reply go.mod:9]: @luckymrwang It is still v0.23.2, maybe you missed the change.

File: pkg/ravencontroller/controllers/gateway_controller.go (hunk 3 above)
Comment (DrmagicE) [reply pkg/ravencontroller/controllers/gateway_controller.go:264]: @njucjc good idea, agree.

**Comment by DrmagicE** [comment]
lgtm

**Comment by DrmagicE** [comment]
@luckymrwang  Please fix the conflict.

File: pkg/ravencontroller/controllers/gateway_controller.go (hunk 3 above)
Comment (njucjc) [reply pkg/ravencontroller/controllers/gateway_controller.go:264]: > 

@luckymrwang Can you change it together in this PR?

**Comment by luckymrwang** [comment]
> @luckymrwang Please fix the conflict.

@DrmagicE The latest code has been commited. Conflict has been fixed.

**Comment by DrmagicE** [comment]
@luckymrwang Please remove the binary file `raven-controller-manager`

**Comment by luckymrwang** [comment]
> @DrmagicE The latest code has been commited.

@DrmagicE The latest code has been commited.

**Comment by luckymrwang** [comment]
> @luckymrwang Please remove the binary file `raven-controller-manager`

@DrmagicE The latest code has been commited.

File: go.mod (hunk 1 above)
Comment (njucjc) [reply go.mod:9]: github.com/projectcalico/api@v0.0.0-20220505235232-ce7a5122e146  requires k8s.io/api@v0.23.2.

repalce v0.22.3 to v0.23.2 will generate `rest.HTTPClientFor` in clientset (`make generate`) which is not supported in v0.22.3

**Pull request merged**